### PR TITLE
[CBRD-21691] Fix the comparison of jsons

### PR DESCRIPTION
**Pull request by paulgheorghecristian**
http://jira.cubrid.org/browse/CBRD-21691

-Create a common function for least and greatest for client and server

We don't really agree with MySQL because they don't abide to the standard. I suggest we keep the same behavior as other types, such as SET.
meaning we return DB_UNK only when total_order is false and either one of them in NULL. otherwise, we compare comparable types or string compare uncomparables (such as json_array)

### CODE REVIEW

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/parser/type_checking.c (3 incorrect indentation(s) found)
```
 17468 
 17479       error = db_least_or_greatest (arg1, arg2, result, false);
 17485 
```
#### File: src/query/arithmetic.c (1 incorrect indentation(s) found)
```
  5352       error_code = er_errid ();
```
#### File: src/query/fetch.c (13 incorrect indentation(s) found)
```
  3177 	int error;
  3178 	TP_DOMAIN *target_domain;
  3179 
  3181 	if (error != NO_ERROR)
  3182 	  {
  3183 	    goto error;
  3184 	  }
  3209 	TP_DOMAIN *target_domain;
  3210 
  3212 	if (error != NO_ERROR)
  3213 	  {
  3214 	    goto error;
  3215 	  }
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**
--------------
**Please check error codes in src/base/error_code.h and src/compat/dbi_compat.h

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/object/object_primitive.c (4 incorrect indentation(s) found)
```
 17655 	  char *str1 = NULL, *str2 = NULL;
 17656 
 17657 	  str1 = db_json_get_raw_json_body_from_document (doc1);
 17658 	  str2 = db_json_get_raw_json_body_from_document (doc2);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/object/object_primitive.c (4 incorrect indentation(s) found)
```
 17661 	  char *str1 = NULL, *str2 = NULL;
 17662 
 17663 	  str1 = db_json_get_raw_json_body_from_document (doc1);
 17664 	  str2 = db_json_get_raw_json_body_from_document (doc2);
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/compat/db_macro.c [hunk 1]
```diff
@@ -6913,3 +6913,42 @@ db_init_db_json_pointers (DB_JSON * val)
...
+  JSON_DOC *doc = DB_GET_JSON_DOCUMENT (src);
+
+  assert (doc != NULL);
+
+  switch (db_json_get_type (doc))
+    {
+    case DB_JSON_STRING:
+      {
+	const char *str = db_json_get_string_from_document (doc, false);
+	DB_MAKE_STRING (dest, str);
```
Comment (ribram) [inline src/compat/db_macro.c]: check error from DB_MAKE_STRING

File: src/object/object_primitive.c [hunk 2]
```diff
@@ -17577,50 +17579,126 @@ mr_data_cmpdisk_json (void *mem1, void *mem2, TP_DOMAIN * domain, int do_coercio
       first += OR_BYTE_SIZE;
       second += OR_BYTE_SIZE;
 
-      return (strcmp (first, second) == 0) ? DB_EQ : DB_UNK;
+      first_json_body = db_private_strdup (NULL, first);
```
Comment (ribram) [inline src/object/object_primitive.c]: 1. Too much memory duplication. In normal circumstance we shouldn't have to duplicate bodies, but I know we have to clear document and body is automatically cleared. I think we should add the possibility of clearing body, document and schema separately. In a different issue, not in this patch.
2. If first or second is compressed, both strings are uncompressed. I think we have to treat each separately.

File: src/object/object_domain.c [hunk 3]
```diff
@@ -11322,7 +11322,7 @@ tp_value_compare_with_error (const DB_VALUE * value1, const DB_VALUE * value2, i
 		  result = (*(pr_type->cmpval)) (v1, v2, do_coercion, total_order, NULL, common_coll);
 		}
 
-	      if (result == DB_UNK)
+	      if (result == DB_UNK || result == DB_NE)
```
Comment (ribram) [inline src/object/object_domain.c]: Why is DB_NE case added here?

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/parser/type_checking.c (2 incorrect indentation(s) found)
```
 17465 	  ASSERT_ERROR ();
 17483 	  ASSERT_ERROR ();
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

**Comment by cubridci** [comment]
Build comment file: 
--------------
#### File: src/parser/type_checking.c (2 incorrect indentation(s) found)
```
 17465 	  ASSERT_ERROR ();
 17483 	  ASSERT_ERROR ();
```
--------------
**Please run 'indent -l120 -lc120' to correct the indentation**

--------------

File: src/object/object_primitive.c [hunk 4]
```diff
@@ -17601,26 +17591,113 @@ mr_data_cmpdisk_json (void *mem1, void *mem2, TP_DOMAIN * domain, int do_coercio
...
+
+  db_make_json (&json1, first_json_body, doc1, true);
+  db_make_json (&json2, second_json_body, doc2, true);
+
+  res = mr_cmpval_json (&json1, &json2, do_coercion, total_order, 0, 0);
 
 cleanup:
-  db_private_free (NULL, first_json_body);
-  db_private_free (NULL, second_json_body);
+  pr_clear_value (&json1);
```
Comment (eseokoh) [inline src/object/object_primitive.c]: please notice that error handling of `pr_get_compressed_data_from_buffer` is bad.
At the moment of jumping, `json1` and `json2` might be uninited and `first_json_body` and `second_json_body` might be leaked.

please also notice `strc` is unused.

File: src/compat/db_json.hpp [hunk 5]
```diff
@@ -39,15 +39,19 @@ typedef void JSON_VALIDATOR;
 #include <functional>
 #include "thread.h"
 
+/*
+ * these also double as type precedence
+ * INT and DOUBLE actually have the same precedence
+*/
```
Comment (eseokoh) [inline src/compat/db_json.hpp]: Can you shift it to right?

File: src/object/object_primitive.c [hunk 6]
```diff
@@ -17601,26 +17591,113 @@ mr_data_cmpdisk_json (void *mem1, void *mem2, TP_DOMAIN * domain, int do_coercio
...
 static DB_VALUE_COMPARE_RESULT
 mr_cmpval_json (DB_VALUE * value1, DB_VALUE * value2, int do_coercion, int total_order, int *start_colp, int collation)
 {
-  bool res = db_json_are_docs_equal (DB_GET_JSON_DOCUMENT (value1), DB_GET_JSON_DOCUMENT (value2));
+  JSON_DOC *doc1 = NULL, *doc2 = NULL;
+  DB_JSON_TYPE type1 = DB_JSON_UNKNOWN;
+  DB_JSON_TYPE type2 = DB_JSON_UNKNOWN;
+  DB_VALUE scalar_value1, scalar_value2;
+  DB_VALUE_COMPARE_RESULT cmp_result;
+  bool can_compare, is_value1_null = true, is_value2_null = true;
```
Comment (eseokoh) [inline src/object/object_primitive.c]: `can_compare` is unused.

File: src/compat/db_json.hpp (hunk 5 above)
Comment (paulgheorghecristian) [reply src/compat/db_json.hpp]: I'll remove it altogether because we no longer rely on type precedence.

**Pull request merged**